ipa-core/src/secret_sharing/mod.rs

```diff
@@ -57,13 +109,140 @@ pub trait Block: Sized + Copy + Debug {
+// SharedValue: Vectorizable<1>`, but doing both creates a cycle. (Similarly for
+// `FieldVectorizable` / `Field`.)
+//
+// Although it is somewhat unnatural, we choose to do the latter, because it allows existing
+// high-level protocols unaware of vectorization to call vectorized versions of core protocols (with
+// width of 1) without updating all of the trait bounds. This does mean that the trait definitions
+// do not prevent implementing `Vectorizable` for something that is not a `SharedValue`, but please
+// don't do that.
+
+/// Trait for `SharedValue`s supporting operations on `N`-wide vectors.
+pub trait Vectorizable<const N: usize>: Sized {
+    type Array: SharedValueArray<Self>;
+}
+
+/// Trait for `Field`s supporting operations on `N`-wide vectors.
+///
+/// We would like `F` to be `FieldVectorizable` if it satisfies all of the following:
+///  1. `F: Field`.
+///  2. `<F as Vectorizable<N>>::Array: FieldArray<Self>`. Rust does not support expressing a
+///     constraint on a super-trait's associated type directly. Instead, this effect is achieved
+///     by constraining the `ArrayAlias` associated type and then constraining that
+///     `Vectorizable::Array == FieldVectorizable::ArrayAlias` where necessary (e.g. in the
+///     definition and blanket impl of the `FieldSimd` trait. We call it `ArrayAlias` instead of
+///     `Array` so that references to the `Array` associated type do not require qualification
+///     with a trait name.
+///  3. `F: Vectorizable<N>`. This is implied by the previous two, because `FieldArray`
+///     is a sub-trait of `SharedValueArray`.
+pub trait FieldVectorizable<const N: usize>: SharedValue + Sized {
+    type ArrayAlias: FieldArray<Self>;
+}
+
+// We could define a `SharedValueSimd` trait that is the analog of this for `SharedValue`s, but
+// there are not currently any protocols that need it.
+pub trait FieldSimd<const N: usize>:
+    Field + Vectorizable<N, Array = <Self as FieldVectorizable<N>>::ArrayAlias> + FieldVectorizable<N>
+{
+}
+
+// Portions of the implementation treat non-vectorized operations as a vector with `N = 1`. This
+// blanket impl (and the fact that `F: Field` is the only trait bound) is important in allowing code
+// that writes `F: Field` to continue working without modification.
+impl<F: Field> FieldSimd<1> for F {}
+
+// Supported vectorizations
+
+impl FieldSimd<32> for Fp32BitPrime {}
+
+pub trait SharedValueArray<V>:
+    Clone
+    + Eq
+    + Debug
+    + Send
+    + Sync
+    + Sized
+    + TryFrom<Vec<V>, Error = ()>
+    + FromIterator<V>
+    + IntoIterator<Item = V>
+    + Add<Self, Output = Self>
+    + for<'a> Add<&'a Self, Output = Self>
+    + AddAssign<Self>
+    + for<'a> AddAssign<&'a Self>
+    + Neg<Output = Self>
+    + Sub<Self, Output = Self>
+    + for<'a> Sub<&'a Self, Output = Self>
+    + SubAssign<Self>
+    + for<'a> SubAssign<&'a Self>
+    + Message
```

Comment:
I would imagine the reverse to be a better model to avoid cyclic dependencies. Secret sharing is a core module and infrastructure stuff already depends on it.

Reply:
It's an overlapping trait impl problem. There can't be wildcard `Message` impls for both `V: SharedValue` and `A: SharedValueArray`. (There's also a problem with the type parameter of `SharedValueArray` being unconstrained in that impl, but that might be resolvable by changing it to an associated type.)

One strategy would be to implement `Message` for anything that is serializable, i.e. `impl<T: Debug + Send + Serializable + 'static> Message for T {}`. That makes it easier to inadvertently create a channel for some type that wasn't really meant to be sent over the network -- not sure how serious a concern that is.

Another possibility would be moving the `Message` trait down to the bottom layer and implementing it explicitly on each applicable type (fields, BAs, etc.), like `Vectorizable`.

Or maybe implementing `SharedValue` for `SharedValueArray`, if we're okay getting rid of `trait Block` and associated type `SharedValue::Store`, neither of which is used very extensively.

Reply:
do you mind if I just create another trait: `Sendable` and implement it for `SharedValue` and `SharedValueArray`? Then we can implement `Message` just for `Sendable` and be done with it. 

Any other hierarchy I feel may be confusing

Reply:
I'm not sure that is sufficient? `SharedValue` and `SharedValueArray` are traits, so `Sendable` would need to be implemented for `T: SharedValue(Array)?` , and the compiler will reject them as potentially overlapping like it does with `Message`.

If the idea is to add `Sendable` to core rather than moving `Message` to core, then maybe implementing it explicitly on fields, BAs, etc. rather than for a type parameter helps?

Reply:
I did some experiments with `Sendable` here: https://github.com/private-attribution/ipa/commit/f1a3b5205e5331b1f4fe26bbc1e070a8fc5fb374

I think we can make it work relatively easy but the purpose of `Message` trait in this case becomes vague

Reply:
That seems like a reasonable approach to me. I don't think `Sendable` and `Message` are distinguished by purpose, but one is used in the protocol code and the other is used in infrastructure code. I tweaked your commit a little bit and added it to this branch.